sys/include/vfs.h

```diff
@@ -0,0 +1,382 @@
+     * the file is still considered closed.
+     *
+     * @return 0 on success
+     * @return <0 on error, the file is considered closed anyway
+     */
+    int (*close) (vfs_file_t *filp);
+
+    /**
+     * @brief Get status of an open file
+     *
+     * @param[in]  filp     pointer to open file
+     * @param[out] buf      pointer to stat struct to fill
+     *
+     * @return 0 on success
+     * @return <0 on error
+     */
+    int (*fstat) (vfs_file_t *filp, struct stat *buf);
+
+    /**
+     * @brief Seek to position in file
+     *
+     * @p whence determines the function of the seek and should be set to one of
+     * the following values:
+     *
+     *  - @c SEEK_SET: Seek to absolute offset @p off
+     *  - @c SEEK_CUR: Seek to current location + @p off
+     *  - @c SEEK_END: Seek to end of file + @p off
+     *
+     * @param[in]  filp     pointer to open file
+     * @param[in]  off      seek offset
+     * @param[in]  whence   determines the seek method, see detailed description
+     *
+     * @return the new seek location in the file on success
+     * @return <0 on error
+     */
+    off_t (*lseek) (vfs_file_t *filp, off_t off, int whence);
+
+    /**
+     * @brief Attempt to open a file in the file system at rel_path
+     *
+     * A file system driver should perform the necessary checks for file
+     * existence etc in this function.
+     *
+     * The VFS layer will initialize the contents of @p *filp so that
+     * @c filp->f_op points to the mounted file system's @c vfs_file_ops_t.
+     * @c filp->private_data will be initialized to NULL, @c filp->pos will be set to 0.
+     *
+     * @note @p name is an absolute path inside the file system, @p abs_path is
+     * the path to the file in the VFS, example: @p abs_path = "/mnt/hd/foo/bar", @p name = "/foo/bar"
+     *
+     * @note @p name and @p abs_path may point to different locations within the
+     * same const char array and the strings may overlap
+     *
+     * @param[in]  filp     pointer to open file
+     * @param[in]  name     null-terminated name of the file to open, relative to the file system root, including a leading slash
+     * @param[in]  flags    flags for opening, see man 2 open, man 3p open
+     * @param[in]  mode     mode for creating a new file, see man 2 open, man 3p open
+     * @param[in]  abs_path null-terminated name of the file to open, relative to the VFS root ("/")
+     *
+     * @return 0 on success
+     * @return <0 on error
+     */
+    int (*open) (vfs_file_t *filp, const char *name, int flags, int mode, const char *abs_path);
+
+    /**
+     * @brief Read bytes from an open file
+     *
+     * @param[in]  filp     pointer to open file
+     * @param[in]  dest     pointer to destination buffer
+     * @param[in]  nbytes   maximum number of bytes to read
+     *
+     * @return number of bytes read on success
+     * @return <0 on error
```

Comment:
What would `return 0` mean? End of file **OR** try again **OR** query some function which of both?

I'd prohibit the result `0` to make things clear. Return `-ENODATA` for EOF, and `-EAGAIN` for try again.

I am not sure about `ENODATA`, though. Maybe `ERANGE`?